File: Algorithm/leetcode_684.go
/**
 * File              : leetcode_684.go
 * Author            : Andy
 * Date              : 2022.08.23
 * Last Modified Date: 2022.08.23
 * Last Modified By  : Andy
 */

// 链接：https://leetcode.cn/problems/redundant-connection/solution/684-rong-yu-lian-jie-bing-cha-ji-ji-chu-eartc/

package findredundantconnect

var (
	n      = 1001
	father = make([]int, 1001)
)

func initialize() {
	for i := 0; i < n; i++ {
		father[i] = i
	}
}

func find(u int) int {
	if u == father[u] {
		return u
	}
	father[u] = find(father[u])
	return father[u]
}

func join(u, v int) {
	u, v = find(u), find(v)
	if u == v {
		return
	}

	father[u] = v
}

func same(u, v int) bool {
	return find(u) == find(v)
}

func findRedundantConnection(edges [][]int) []int {
	initialize()

	for i := 0; i < len(edges); i++ {
		if same(edges[i][0], edges[i][1]) {
			return edges[i]
		} else {
			join(edges[i][0], edges[i][1])
		}
	}
	return []int{}
}

File: Algorithm/leetcode_684.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
# File              : leetcode_684.py
# Author            : Andy
# Date              : 2022.08.22
# Last Modified Date: 2022.08.22
# Last Modified By  : Andy


class UnionSet:
    def __init_(self, n):
        self.root = [-1] * (n + 1)

    def find(self, node):
        while self.root[node] != -1:
            node = self.root[node]
        return node

    def connect(self, node1, node2):
        first_root, second_root = self.find(node1), self.find(node2)
        if first_root != second_root:
            self.root[first_root] = second_root

class Solution:
    def findRedundantConnection(self, edges):
        node_set = set()
        for edge in edges:
            node_set.add(node[0])
            node_set.add(node[1])
        us = UnionSet(len(node_set))

        for e in edges:
            first, second = us.find(e[0]), us.find(e[1])
            if first != second:
                us.connect(first, second)
            else:
                return e
